File: algorithms/3_hard/C/00630_course_schedule_iii/heap.cpp
class Solution {
public:
    static int scheduleCourse(vector<vector<int>>& courses) {
        ranges::sort(courses, [](const auto& a, const auto& b) {
            return a[1] < b[1];
        });
        auto heap = priority_queue<int>();
        int sum = 0;

        for (const auto& c : courses) {
            sum += c[0];
            const int max_allowed = c[1];
            heap.push(c[0]);

            if (sum > max_allowed) {
                sum -= heap.top();
                heap.pop();
            }
        }

        return static_cast<int>(std::ssize(heap));
    }
};

File: algorithms/3_hard/C/00630_course_schedule_iii/heap.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn schedule_course(mut courses: Vec<Vec<i32>>) -> i32 {
        courses.sort_unstable_by(|a, b| a[1].cmp(&b[1]));
        let mut heap: BinaryHeap<i32> = BinaryHeap::with_capacity(courses.len());
        let mut sum = 0_i32;

        for c in courses {
            sum += c[0];
            let max_allowed = c[1];
            heap.push(c[0]);

            if sum > max_allowed {
                sum -= heap.pop().unwrap();
            }
        }

        heap.len().try_into().unwrap()
    }
}
